src/merkle_tree.rs

```diff
@@ -1,127 +1,78 @@
-    // It's an option type because maybe the child does not
-    // exist.
-    left_child: Option<Position>,
-    right_child: Option<Position>,
-}
-
-impl Leaf {
-    pub fn new(
-        hash: [u8; 32],
-        left_child: Option<Position>,
-        right_child: Option<Position>,
-    ) -> Leaf {
-        Leaf {
-            hash,
-            left_child,
-            right_child,
-        }
-    }
-}
-
-pub fn hash(positions: [Position; 2], leaves: &[&Leaf; 2]) -> Leaf {
+pub fn hash(leaves: &[HashedData]) -> HashedData {
     let new_hash = leaves
         .iter()
         .fold(Sha3_256::new(), |mut acc, a| {
-            let value = a.hash;
+            let value = a;
             acc.update(value);
             acc
         })
-        .finalize();
+        .finalize().into();
 
-    Leaf {
-        hash: new_hash.into(),
-        right_child: Some(positions[1]),
-        left_child: Some(positions[0]),
-    }
+    new_hash
 }
 
 pub struct MerkleTree {
-    leaves: Vec<Leaf>,
+    leaves: Vec<Vec<HashedData>>,
 }
 
 impl MerkleTree {
     pub fn new<T: std::convert::AsRef<[u8]>>(data: &[T]) -> MerkleTree {
-
-
-        // The first time the "latest leaves" are created from the
-        // node passed as parameters.
-        let mut latest_leaves: Vec<(Position, Leaf)> = data
+        let initial_blocks: Vec<HashedData> = data
             .iter()
             .map(|value| {
-                Sha3_256::digest(value)
+                Sha3_256::digest(value).into()
             })
-            .map(|hash| Leaf::new(hash.into(), None, None))
-            .enumerate()
             .collect();
 
-        // We will extend the trees with the latest leaves every time.
-        let mut tree: Vec<(Position, Leaf)> = vec![];
+        let mut tree: Vec<Vec<HashedData>> = vec![];
 
-        // I hate this so much
+        let mut new_leaves = initial_blocks;
+        tree.push(new_leaves.clone());
         loop {
-	  tree.extend(latest_leaves.clone());
-	  latest_leaves = Self::add_children_leaves(latest_leaves, tree.len());
-	  if latest_leaves.len() == 1 {
-	      tree.extend(latest_leaves.clone());
+
+	  new_leaves = new_leaves
+	      .chunks(2)
+	      .map(|left_n_right|
+		 {
+		     if left_n_right.get(1).is_none() {
+		        // This is the case where we have an uneven
+		        // amount of nodes
+		         hash(&[left_n_right[0]])
+                             // [&left_n_right[0], &left_n_right[0]]
+
+		     } else {
+		         hash(&[left_n_right[0], left_n_right[1]])
+		     }
+		 }
+	      )
+	      .collect();
+	  tree.push(new_leaves.clone());
+	  if new_leaves.len() == 1 {
 	      break;
 	  }
         }
 
-        println!("");
-        for leaf in tree.clone() {
-	  println!("{:?}", leaf);
+        println!("HERE");
+        println!("{:?}", tree);
+        for level in tree.clone() {
+	  // for node in level {
+	  //     print!("{:?}", node);
+	  // }
+	  println!("{}", level.len())
         }
```

Comment:
You can create a function to perform printing. Or maybe implement a Display trait.

Reply:
Thanks for the tip! I'll implement it later on however